snow/networking/handler/health_test.go

```diff
@@ -0,0 +1,138 @@
+	"github.com/ava-labs/avalanchego/utils/math/meter"
+	"github.com/ava-labs/avalanchego/utils/resource"
+	"github.com/ava-labs/avalanchego/utils/set"
+)
+
+func TestHealthCheckSubnet(t *testing.T) {
+	tests := map[string]struct {
+		minStake float64
+	}{
+		"default min stake": {
+			minStake: constants.DefaultMinConnectedStake,
+		},
+		"custom min stake": {
+			minStake: 0.40,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			require := require.New(t)
+
+			ctx := snow.DefaultConsensusContextTest()
+
+			vdrs := validators.NewSet()
+
+			resourceTracker, err := tracker.NewResourceTracker(
+				prometheus.NewRegistry(),
+				resource.NoUsage,
+				meter.ContinuousFactory{},
+				time.Second,
+			)
+			require.NoError(err)
+
+			peerTracker := commontracker.NewPeers()
+			vdrs.RegisterCallbackListener(peerTracker)
+			handlerIntf, err := New(
+				ctx,
+				vdrs,
+				nil,
+				time.Second,
+				testThreadPoolSize,
+				resourceTracker,
+				validators.UnhandledSubnetConnector,
+				subnets.New(ctx.NodeID, subnets.Config{}),
+				peerTracker,
+			)
+
+			bootstrapper := &common.BootstrapperTest{
+				BootstrapableTest: common.BootstrapableTest{
+					T: t,
+				},
+				EngineTest: common.EngineTest{
+					T: t,
+				},
+			}
+			bootstrapper.Default(false)
+
+			engine := &common.EngineTest{T: t}
+			engine.Default(false)
+			engine.ContextF = func() *snow.ConsensusContext {
+				return ctx
+			}
+
+			handlerIntf.SetEngineManager(&EngineManager{
+				Snowman: &Engine{
+					Bootstrapper: bootstrapper,
+					Consensus:    engine,
+				},
+			})
+
+			ctx.State.Set(snow.EngineState{
+				Type:  p2p.EngineType_ENGINE_TYPE_SNOWMAN,
+				State: snow.NormalOp, // assumed bootstrap is done
+			})
+
+			bootstrapper.StartF = func(context.Context, uint32) error {
+				return nil
+			}
+
+			handlerIntf.Start(context.Background(), false)
+
+			testVdrCount := 4
+			vdrIDs := set.NewSet[ids.NodeID](testVdrCount)
+			for i := 0; i < testVdrCount; i++ {
+				vdrID := ids.GenerateTestNodeID()
+				require.NoError(err)
+				err := vdrs.Add(vdrID, nil, ids.Empty, 100)
+				require.NoError(err)
+				vdrIDs.Add(vdrID)
+			}
+
+			ctx.MinPercentConnectedStakeHealthy = test.minStake
+
+			for index, vdr := range vdrs.List() {
+				err := peerTracker.Connected(context.Background(), vdr.NodeID, nil)
+				require.NoError(err)
+				details, err := handlerIntf.HealthCheck(context.Background())
+				connectedPerc := float64(index+1) / float64(testVdrCount)
+				if connectedPerc >= test.minStake {
+					require.NoError(err)
+				} else {
+					expectedDetails := map[string]float64{
+						"percentConnected": connectedPerc,
+					}
+					require.Contains(fmt.Sprint(details), fmt.Sprint(expectedDetails))
+					err := fmt.Errorf("connected to %f%% of network stake; should be connected to at least %f%%",
+						connectedPerc*100,
+						test.minStake*100,
+					)
+					require.Contains(err.Error(), err.Error())
```

Comment:
cc: @dhrubabasu

Reply:
I feel like the best thing we can do here is use `require.ErrorIs` for the first error and add a TODO for after go1.20 to test for the second error.

Reply:
added a new error type